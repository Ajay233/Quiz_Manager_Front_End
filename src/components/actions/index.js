export const setToTopButton = (show=false) => {
  return {
    type: "SHOW_TO_TOP_BUTTON",
    payload: show
  }
}

export const setUrl = (url="") => {
  return {
    type: "SET_URL",
    payload: url
  }
}

export const enableLightMode = () => {
  return {
    type: "SET_THEME",
    payload: false
  }
}

export const enableDarkMode = () => {
  return {
    type: "SET_THEME",
    payload: true
  }
}

export const setLoaderState = (show=false, message="", label="") => {
  return {
    type: "SET_LOADER_STATE",
    payload: {
      show: show,
      message: message,
      label: label
    }
  }
}
