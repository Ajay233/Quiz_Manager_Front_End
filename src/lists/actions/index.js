import { get } from '../../axiosRequests/requests'
import { sessionExpired } from '../../utils/session'

export const getCategories = (endpoint, jwt) => {
  return (dispatch) => {
    get(endpoint, jwt).then((response) => {
      dispatch({
        type: "SET_CATEGORIES",
        payload: response.data
      })
    }).catch((error) => {
      console.log(error.response)
      if(error.response.status === 403){
        sessionExpired(dispatch);
      }
    })
  }
}
