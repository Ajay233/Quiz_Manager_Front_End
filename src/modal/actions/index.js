export const showModal = () => {
  return {
    type: "SHOW_MODAL",
    payload: true
  }
}

export const hideModal = () => {
  return {
    type: "HIDE_MODAL",
    payload: false
  }
}
