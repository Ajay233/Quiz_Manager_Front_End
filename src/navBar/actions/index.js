export const setDropDownState = (show=false) => {
  return {
    type: "SET_DROP_DOWN_STATE",
    payload: show
  }
}
