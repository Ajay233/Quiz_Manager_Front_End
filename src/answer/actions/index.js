import { getUsingParams, post, put, del } from '../../axiosRequests/requests'
import { setNotification } from '../../notifications/actions'
import { hideModal } from '../../modal/actions'
import { sessionExpired } from '../../utils/session'
import history from '../../history'

// create
export const addAnswer = (data, jwt) => {
  return (dispatch) => {
    return post("answer/create", [data], jwt).then((response) => {
      dispatch({
        type: "ADD_ANSWER",
        payload: response.data[0]
      })
      history.push("/editQuestion");
      dispatch(setNotification("Answer created", "success", true));
    }).catch((error) => {
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        console.log(error.response);
        dispatch(setNotification("Error - unable to create answer", "error", true));
      }
    });
  }
}

// read
export const getAnswers = (param, jwt) => {
  return (dispatch) => {
    return getUsingParams("answer/findByQuestionId", param, jwt).then((response) => {
      dispatch({
        type: "SET_ANSWERS",
        payload: response.data
      });
    }).catch((error) => {
      dispatch(setNotification("Error retrieving answers", "error", true));
    });
  }
}

// update (there was a call to setCurrentAnswer but have removed this as it's not needed)
export const updateAnswer = (body, jwt) => {
  return (dispatch) => {
    return put("answer/update", [body], jwt).then((response) => {
      dispatch({
        type: "UPDATE_ANSWER",
        payload: response.data[0]
      })
      history.push("/editQuestion");
      dispatch(setNotification("Answer updated", "success", true));
    }).catch((error) => {
      console.log(error.response);
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        dispatch(setNotification("Error - unable to update answer", "error", true));
      }
    });
  }
}

// delete
export const deleteAnswer = (config, jwt) => {
  return (dispatch) => {
    return del("answer/delete", config, jwt).then((response) => {
      dispatch(hideModal())
      dispatch({
        type: "DELETE_ANSWER",
        payload: config.data[0]
      })
      dispatch(setNotification("Answer deleted", "success", true));
    }).catch((error) => {
      console.log(error.response);
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        dispatch(hideModal())
        dispatch(setNotification("Error - Unable to delete this answer", "error", true))
      }
    });
  }
}

export const setCurrentAnswer = (answer) => {
  return {
    type: "SET_CURRENT_ANSWER",
    payload: answer
  }
}
