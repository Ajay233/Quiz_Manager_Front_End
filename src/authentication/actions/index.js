import { post } from '../../axiosRequests/requests'
import { setNotification } from '../../notifications/actions'
import history from '../../history'

export const setUser = (endpoint, loginDetails) => {

  return (dispatch) => {

    post(endpoint, loginDetails).then((response) => {
      dispatch({
        type: "SET_USER_LOGGED_IN",
        payload: {
          id: response.data.user.id,
          forename: response.data.user.forename,
          surname: response.data.user.surname,
          email: response.data.user.email,
          permission: response.data.user.permission,
          verified: response.data.user.verified,
          jwt: response.data.jwt,
          loggedIn: true,
          response: response
        }
      })
      history.push("/")
      dispatch(setNotification(`Welcome back ${response.data.user.forename}`, "loginSuccess", true))
    }).catch((error) => {
      if(error.response.data === "NOT VERIFIED"){
        const verifyMsg = "Your email has not yet been veirified.  You will need to verify your email before you can log in"
        dispatch(setNotification(verifyMsg, "warning", true))
      } else {
        dispatch({
          type: "SET_USER_LOGGED_IN",
          payload: {
            error: error.response
          }
        });
        dispatch(setNotification("The username or password you entered was incorrect", "error", true));
      }
    })
  }
}

export const logOutUser = () => {
  return {
    type: "LOG_OUT_USER",
    payload: {
      forename: "",
      surname: "",
      email: "",
      permission: "",
      verified: "",
      jwt: "",
      loggedIn: false
    }
  }
}

export const setVerficationProcess = (endpoint, verificationDetails) => {

  return (dispatch) => {

    post(endpoint, verificationDetails).then((response) => {
      dispatch({
        type: "SET_VERIFY_PROCESS_STATUS",
        payload: {
          completionStatus: "completed",
          token: "",
          error: {}
        }
      })
    }).catch((error) => {
      dispatch({
        type: "SET_VERIFY_PROCESS_STATUS",
        payload: {
          completionStatus: "not verified",
          token: verificationDetails.token,
          error: error.response
        }
      })
    })
  }
}

export const setRedirect = (status) => {
  return {
    type: "SET_REDIRECT",
    payload: {
      status: status
    }
  }
}

export const logOut = () => {
  return {
    type: "RESET_APP"
  }
}

// export const getUsers = (endpoint, header) => {
//   return {
//     type: "GET_USER",
//     payload: {
//       userId
//   }
// }
