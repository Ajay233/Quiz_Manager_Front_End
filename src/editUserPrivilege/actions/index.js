import { getUsingParams } from '../../axiosRequests/requests'
import { setNotification } from '../../notifications/actions'
import { sessionExpired } from '../../utils/session'

export const setUserResults = (param, jwt) => {
  return (dispatch) => {
    return getUsingParams("users/findByEmail", param, jwt).then((response) => {
      dispatch({
        type: "SET_USER_RESULTS",
        payload: response.data
      })
    }).catch((error) => {
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        const msg = error.response.data;
        dispatch(setNotification(msg, "error", true));
        console.log(error.response)
      }
    });
  }
}

export const clearUserResults = () => {
  return {
    type: "CLEAR_USER_RESULTS",
    payload: {}
  }
}
