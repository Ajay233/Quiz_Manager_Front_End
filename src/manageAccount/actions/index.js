import { put } from '../../axiosRequests/requests'
import { setUserData } from '../../authentication/actions'
import { setNotification } from '../../notifications/actions'
import { sessionExpired } from '../../utils/session'

export const updateUser = (data, jwt) => {
  const emailChangeMsg = "Profile information updated - Your email was changed so your access has been restriced until your email is verified"
  const errorMsg = "Error updating your profile data, please check the details you provided and try again";
  const emailTaken = "An account with that email already exists"
  return (dispatch) => {
    return put('users/update', data, jwt).then((response) => {
      if(response.data.jwt === undefined){
        dispatch(setUserData(response.data, jwt))
        dispatch(setNotification("Profile information updated", "success", true))
      } else {
        dispatch(setUserData(response.data.user, response.data.jwt))
        dispatch(setNotification(emailChangeMsg, "success", true))
      }
    }).catch((error) => {
      console.log(error)
      console.log(error.response)
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else if(error.response.data === emailTaken){
        dispatch(setNotification(emailTaken, "error", true))
      } else {
        dispatch(setNotification(errorMsg, "error", true))
      }
    });
  }
}
