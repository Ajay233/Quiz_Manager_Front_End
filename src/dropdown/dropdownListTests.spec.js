import React from 'react'
import { Provider } from 'react-redux'
import { Router } from 'react-router-dom'
import DropdownList from './dropdownList'
import { shallow, mount } from 'enzyme'
import renderer from 'react-test-renderer'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import history from '../history'

import { setNotification } from '../notifications/actions'
import { logOut } from '../authentication/actions'
import { getAllQuizes } from '../quizSearch/actions'

jest.mock("../history")
jest.mock("../authentication/actions")
jest.mock("../notifications/actions")
jest.mock("../quizSearch/actions")

const middlewares = [thunk]
const mockStore = configureStore(middlewares)

describe("DropdownList", () => {

  let store;
  let wrapper;

  beforeEach(() => {

    store = mockStore({
      userData: {
        permission: "USER",
        loggedIn: true
      }
    })

    wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <DropdownList
            logOut={logOut}
            setNotification={setNotification}
            getAllQuizes={getAllQuizes}
          />
        </Router>
      </Provider>
    )
  })

  it("should call action creators and push a new location pathname on logout", () => {
    wrapper.find('button').simulate('click')
    expect(logOut).toHaveBeenCalledTimes(1)
    expect(setNotification).toHaveBeenCalledTimes(1)
    expect(history.push).toHaveBeenCalledTimes(1)

  })

  it("should call getAllQuizes when browse all is clicked", () => {
    wrapper.find('Link').at(1).simulate('click')
    expect(getAllQuizes).toHaveBeenCalledTimes(1)
  })

  it("should render only login and create account if user is not logged in", () => {
    store = mockStore({
      userData: {
        permission: "USER",
        loggedIn: false
      }
    })

    const component = renderer.create(
      <Provider store={store}>
        <Router history={history}>
          <DropdownList
            logOut={logOut}
            setNotification={setNotification}
            getAllQuizes={getAllQuizes}
          />
        </Router>
      </Provider>
    )

    expect(component).toMatchSnapshot()
  })

  it("should render only USER options for user with USER permission", () => {
    store = mockStore({
      userData: {
        permission: "USER",
        loggedIn: true
      }
    })

    const component = renderer.create(
      <Provider store={store}>
        <Router history={history}>
          <DropdownList
            logOut={logOut}
            setNotification={setNotification}
            getAllQuizes={getAllQuizes}
          />
        </Router>
      </Provider>
    )

    expect(component).toMatchSnapshot()
  })

  it("should render only READ-ONLY options for user with READ-ONLY permission", () => {
    store = mockStore({
      userData: {
        permission: "READ-ONLY",
        loggedIn: true
      }
    })

    const component = renderer.create(
      <Provider store={store}>
        <Router history={history}>
          <DropdownList
            logOut={logOut}
            setNotification={setNotification}
            getAllQuizes={getAllQuizes}
          />
        </Router>
      </Provider>
    )

    expect(component).toMatchSnapshot()
  })

  it("should render only ADMIN options for user with ADMIN permission", () => {
    store = mockStore({
      userData: {
        permission: "ADMIN",
        loggedIn: true
      }
    })

    const component = renderer.create(
      <Provider store={store}>
        <Router history={history}>
          <DropdownList
            logOut={logOut}
            setNotification={setNotification}
            getAllQuizes={getAllQuizes}
          />
        </Router>
      </Provider>
    )

    expect(component).toMatchSnapshot()
  })

})
