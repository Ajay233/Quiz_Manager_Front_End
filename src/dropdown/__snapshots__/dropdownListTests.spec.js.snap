// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownList should render only ADMIN options for user with ADMIN permission 1`] = `
<div
  className="list"
>
  <ul>
    <li>
      <a
        className="links"
        onClick={[Function]}
      >
        Manage my account
      </a>
      <hr />
    </li>
    <li>
      <a
        className="links"
        onClick={[Function]}
      >
        List all Users
      </a>
      <hr />
    </li>
    <li>
      <a
        className="links"
        onClick={[Function]}
      >
        Edit Privileges
      </a>
      <hr />
    </li>
    <li>
      <a
        className="links"
        onClick={[Function]}
      >
        Browse All Quizes
      </a>
      <hr />
    </li>
    <li>
      <a
        className="links"
        onClick={[Function]}
      >
        Quiz search
      </a>
      <hr />
    </li>
    <li>
      <a
        className="links"
        onClick={[Function]}
      >
        Create a Quiz
      </a>
      <hr />
    </li>
    <li>
      <button
        className="linkButton links"
        onClick={[Function]}
      >
        Logout
      </button>
    </li>
  </ul>
</div>
`;

exports[`DropdownList should render only READ-ONLY options for user with READ-ONLY permission 1`] = `
<div
  className="list"
>
  <ul>
    <li>
      <a
        className="links"
        onClick={[Function]}
      >
        Manage my account
      </a>
      <hr />
    </li>
    <li>
      <a
        className="links"
        onClick={[Function]}
      >
        Browse All Quizes
      </a>
      <hr />
    </li>
    <li>
      <a
        className="links"
        onClick={[Function]}
      >
        Quiz search
      </a>
      <hr />
    </li>
    <li>
      <button
        className="linkButton links"
        onClick={[Function]}
      >
        Logout
      </button>
    </li>
  </ul>
</div>
`;

exports[`DropdownList should render only USER options for user with USER permission 1`] = `
<div
  className="list"
>
  <ul>
    <li>
      <a
        className="links"
        onClick={[Function]}
      >
        Manage my account
      </a>
      <hr />
    </li>
    <li>
      <a
        className="links"
        onClick={[Function]}
      >
        Browse All Quizes
      </a>
      <hr />
    </li>
    <li>
      <a
        className="links"
        onClick={[Function]}
      >
        Quiz search
      </a>
      <hr />
    </li>
    <li>
      <button
        className="linkButton links"
        onClick={[Function]}
      >
        Logout
      </button>
    </li>
  </ul>
</div>
`;

exports[`DropdownList should render only login and create account if user is not logged in 1`] = `
<div
  className="list"
>
  <ul>
    <li>
      <a
        className="links"
        onClick={[Function]}
      >
        Login
      </a>
      <hr />
    </li>
    <li>
      <a
        className="links"
        onClick={[Function]}
      >
        Create Account
      </a>
    </li>
  </ul>
</div>
`;
